keyring/backends/Windows/api.py

```diff
@@ -0,0 +1,273 @@
+
+            if field == 'LastWritten':
+                result[field] = value.asdict()
+            elif field == 'Attributes':
+                attr_dict = {}
+                for i in range(result['AttributeCount']):
+                    if value[i]:  # NULL pointer is treated as False by ctypes
+                        attr = value[i].asdict()
+                        attr_dict[attr['Keyword']] = attr
+                result[field] = attr_dict
+            elif field == 'CredentialBlob':
+                blob = string_at(value, getattr(self, 'CredentialBlobSize'))
+                result[field] = blob
+            else:
+                result[field] = value
+
+        return result
+
+
+# endregion
+
+
+# region Win32 API declarations
+
+
+def errcheck(result, func, args):
+    if not result:
+        raise WinError()
+
+
+_CredReadW = windll.advapi32.CredReadW
+_CredReadW.argtypes = [c_wchar_p, DWORD, DWORD, POINTER(POINTER(CREDENTIALW))]
+_CredReadW.restype = BOOL
+_CredReadW.errcheck = errcheck  # type: ignore
+
+_CredWriteW = windll.advapi32.CredWriteW
+_CredWriteW.argtypes = [POINTER(CREDENTIALW), DWORD]
+_CredWriteW.restype = BOOL
+_CredWriteW.errcheck = errcheck  # type: ignore
+
+_CredDeleteW = windll.advapi32.CredDeleteW
+_CredDeleteW.argtypes = [c_wchar_p, DWORD, DWORD]
+_CredDeleteW.restype = BOOL
+_CredDeleteW.errcheck = errcheck  # type: ignore
+
+_CredEnumerateW = windll.advapi32.CredEnumerateW
+_CredEnumerateW.argtypes = [
+    c_wchar_p,
+    DWORD,
+    POINTER(DWORD),
+    POINTER(POINTER(POINTER(CREDENTIALW))),
+]
+_CredEnumerateW.restype = BOOL
+_CredEnumerateW.errcheck = errcheck  # type: ignore
+
+_CredFree = windll.advapi32.CredFree
+_CredFree.argtypes = [LPVOID]
+_CredFree.restype = None
+
+CRED_TYPE_GENERIC = 1
+CRED_PERSIST_SESSION = 1
+CRED_PERSIST_LOCAL_MACHINE = 2
+CRED_PERSIST_ENTERPRISE = 3
+CRED_ENUMERATE_ALL_CREDENTIALS = 1
+CRED_MAX_VALUE_SIZE = 256
+CRED_MAX_ATTRIBUTES = 64
+MAX_PASSWORD_BYTES = CRED_MAX_VALUE_SIZE * CRED_MAX_ATTRIBUTES
+
+
+# endregion
+
+# region API
+
+
+def CredRead(Type, TargetName):
+    pcred = pointer(CREDENTIALW())
+    _CredReadW(TargetName, Type, 0, byref(pcred))
+    credential = pcred.contents.asdict()
+    _CredFree(pcred)
+    return credential
+
+
+def CredReadFromAttributes(Type, TargetName, Credential=None, encoding='utf-8'):
+    if Credential is None:
+        Credential = CredRead(Type, TargetName)
+
+    num_attrs = Credential['AttributeCount']
+    if num_attrs > 0:
+        attrs = Credential['Attributes']
+        accum = b''
+        for i, (key, attr) in enumerate(attrs.items()):
+            expected_key = '{}:{}'.format(ATTRIBUTE_KEYWORD, str(i))
+            if key == expected_key:
+                accum += string_at(attr['Value'], attr['ValueSize'])
+            else:
+                break
+
+        Credential['CredentialBlob'] = accum.decode(encoding)
+        Credential['CredentialBlobSize'] = len(Credential['CredentialBlob'])
```

Comment:
It seems a little odd to me to be utilizing the API structures to embed the altered values, values that wouldn't be allowed in the structure. I think I'd rather see a different approach where the encoding/decoding strategy sits as a layer between the Windows API and the keyring API, and less intertwined with the Windows API.

I can provide an example later.

Reply:
Agreed (see my previous comment for how I got there.)